### PR TITLE
add support of grapheme clusters

### LINKED ISSUE #298: Use unicode-segmentation 1.2 for grapheme cluster boundaries
Version 1.2 of [unicode-segmentation](https://crates.io/crates/unicode-segmentation) has a rope-aware grapheme segmentation API. Change the `Rope::prev_grapheme_offset` and `next_grapheme_offset` functions to use this API rather than just advancing by codepoint.

### PR DESCRIPTION
**Pull request by jmuk**
now next/prev_grapheme_offset works correctly rather than
falling back to codepoints.

fixes #298

### CODE REVIEW

File: rust/rope/src/rope.rs [hunk 1]
```diff
@@ -145,6 +147,51 @@ impl Metric<RopeInfo> for BaseMetric {
...
+    fn to_base_units(_: &String, in_measured_units: usize) -> usize {
+        in_measured_units
+    }
+
+    fn from_base_units(_: &String, in_base_units: usize) -> usize {
+        in_base_units
+    }
+
+    fn is_boundary(s: &String, offset: usize) -> bool {
+        if let Ok(r) = GraphemeCursor::new(offset, s.len(), true).is_boundary(s, 0) {
```
Comment (cmyr) [inline rust/rope/src/rope.rs]: you can call `.ok()` on any `Result<T>` to convert it to an `Option<T>`, discarding the error.

File: rust/rope/src/rope.rs [hunk 2]
```diff
@@ -361,15 +408,20 @@ impl Rope {
         cursor.next::<BaseMetric>()
     }
 
+    pub fn is_grapheme_boundary(&self, offset: usize) -> bool {
+        let mut cursor = Cursor::new(self, offset);
+        cursor.is_boundary::<GraphemeMetric>()
+    }
+
     // graphemes should probably be developed as a cursor-based interface
```
Comment (cmyr) [inline rust/rope/src/rope.rs]: We can get rid of this comment now. :)

File: rust/rope/src/rope.rs [hunk 3]
```diff
@@ -145,6 +147,51 @@ impl Metric<RopeInfo> for BaseMetric {
     }
 }
 
+#[derive(Clone, Copy)]
+pub struct GraphemeMetric(());
```
Comment (cmyr) [inline rust/rope/src/rope.rs]: You can declare a unit struct without any braces: `pub struct GraphemeMetric;`.

**Comment by raphlinus** [comment]
I'm thrilled to see progress on this, and hate to throw cold water on the idea, but unfortunately trying to treat grapheme clusters as a metric doesn't quite work. The underlying reason is that grapheme cluster counting is fundamentally not a monoid homomorphism, as awesome as it would be if that were the case. Here's how to write a test case that will fail:

Create a string of [RIS](https://en.wikipedia.org/wiki/Regional_Indicator_Symbol) code points, let's say `\u{1F1E6}` that doesn't fit in a single leaf node (ie > 1kb). Now test whether the offset at len(first leaf) + 4 bytes is a boundary. The `GraphemeMetric` will always say no, but that would be wrong if the number of code points in the first leaf is odd. If the number of code points in the first leaf is even, edit the rope (usually taking the substring from offset 4 will work, unless the first leaf is small), so that one or the other would fail.

Because the size of graphemes is not bounded, and because boundaries can in the worst case depend on unbounded amounts of context, the boundary detection must be able to handle bringing in an arbitrary number of leaves as pre-context.

Bottom line, this needs to be implemented directly in {prev,next}_grapheme_offset with some attention to leaf boundaries. It is, sadly, not trivial to test - this is one reason I'd like to see some infrastructure for property testing developed.

Thanks for the attempt, and I'm happy to provide guidance to see it through.

**Comment by mqzry** [comment]
I am taking a stab at this. Did I understand it correctly that what is needed is:

- Instantiate a GraphemeCursor
- create a chunk iterators that walks in reverse and normal order to handle `GraphemeIncomplete::PreContext` or `GraphemeIncomplete::NextChunk` results
- Call the right function the cursor for example `is_boundary`
- Handle any `Err(GraphemeIncomplete)` return value properly

**Comment by jmuk** [comment]
Yup, actually I was working on that. Some prototype is working on my local machine although it needs some refinements (sorry for my slow progress).

However, I've found a few issues on `unicode-segmentation` side, which cause `panic`. I'll send a PR to them first. Stay tuned.

**Comment by jmuk** [comment]
https://github.com/unicode-rs/unicode-segmentation/pull/37 here is the PR.

**Comment by jmuk** [comment]
So I pushed my version of the code. Please take a look, if that's okay. I'm fine to hold this until the problem I've seen in unicode-segmentation are fixed.

**Comment by jmuk** [comment]
This PR actually relies on https://github.com/unicode-rs/unicode-segmentation/pull/40 too. I can create workarounds if that turned out unwanted.

**Comment by mqzry** [comment]
What is the status for this PR? All relevant PRs seem to be merged.

**Comment by raphlinus** [comment]
Good question. I think unicode-segmentation needs to be published to crates.io. @Manishearth?

**Comment by Manishearth** [comment]
done!

-Manish Goregaokar

**Comment by raphlinus** [comment]
Thanks @Manishearth! Would you like to resolve the conflicts, @jmuk? If not, we can take care of it.

Incidentally, this will help a lot in dealing with CRLF line endings, right now the cursor can get between the two which causes all kinds of unexpected behavior.

**Comment by jmuk** [comment]
@raphlinus done!

File: rust/rope/src/rope.rs [hunk 4]
```diff
@@ -548,6 +551,75 @@ impl<'a> Cursor<'a, RopeInfo> {
...
+        let mut pos = self.pos();
+        while !l.is_char_boundary(offset) {
+            offset -= 1;
+            pos -= 1;
+        }
+        let leaf_offset = pos - offset;
+        let mut c = GraphemeCursor::new(pos, l.len() + leaf_offset, true);
+        let mut next_boundary = c.next_boundary(&l, leaf_offset);
+        while let Err(GraphemeIncomplete::PreContext(_)) = next_boundary {
+            let (pl, poffset) = self.prev_leaf()?;
```
[the remark this answers is not in the document]
Comment (jmuk) [reply rust/rope/src/rope.rs]: It seems the README had been updated already to require Rust 1.22; as https://github.com/google/xi-editor/commit/3c53388eccb4a38f3f24db2da6960f2a55c827fa

File: rust/rope/src/rope.rs [hunk 5]
```diff
@@ -363,13 +366,13 @@ impl Rope {
 
     // graphemes should probably be developed as a cursor-based interface
```
[the remark this answers is not in the document]
Comment (jmuk) [reply rust/rope/src/rope.rs]: removed!

File: rust/rope/src/rope.rs [hunk 6]
```diff
@@ -548,6 +551,75 @@ impl<'a> Cursor<'a, RopeInfo> {
...
+        let mut c = GraphemeCursor::new(pos, l.len() + leaf_offset, true);
+        let mut next_boundary = c.next_boundary(&l, leaf_offset);
+        while let Err(GraphemeIncomplete::PreContext(_)) = next_boundary {
+            let (pl, poffset) = self.prev_leaf()?;
+            c.provide_context(&pl, self.pos() - poffset);
+            next_boundary = c.next_boundary(&l, leaf_offset);
+        }
+        // GraphemeIncomplete::NextChunk error will not happen, since this cursor thinks the end of
+        // the current leaf as the end of the string. Can a cursor know the total length of
+        // the tree?
```
[the remark this answers is not in the document]
Comment (jmuk) [reply rust/rope/src/rope.rs]: Okay -- yet `cursor.root` isn't visible here in `rope.rs`. I added `total_len()` to `tree.rs`, and it helps to simplify the code; at least I think so. PTAL.

**Comment by raphlinus** [comment]
Many thanks!

**Comment by jmuk** [comment]
Yay!
